compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -390,6 +390,13 @@ public BType checkExpr(BLangExpression expr, SymbolEnv env, BType expType, Diagn
         return resultType;
     }
 
+    private void analyzeNode(BLangNode node, SymbolEnv env) {
+        boolean prevNonErrorLoggingCheck = this.nonErrorLoggingCheck;
+        this.nonErrorLoggingCheck = false;
```

Comment:
Do we know why this fixes the issue when we set `this.nonErrorLoggingCheck = false;`?

Reply:
Seems like this was done to make the lambda function be added in https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L4340 in the initial (non-logging) iteration. IMO, this is not correct. We shouldn't have to special case this.

Everywhere we do a non-error logging check, we eventually do an actual error logging check against some type. For the original issue with the cast, this should happen at https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L4319. And that is when the lambda function should get added.

But during that iteration, we don't seem to get to the point where the lambda function is added because we return at https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L360, which shouldn't be the case because in the previous (non-error logging) iterations we use a clone, so `expr.typeChecked` should be false here. This usually happens when the expression is not properly cloned. IMO, the correct fix is to fix that.

@chiranSachintha, can you please recheck?

Reply:
In here https://github.com/ballerina-platform/ballerina-lang/blob/bab02a8fb3cfd11794e2394e1460b23ec0dfe115/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L8128 we are getting the class definitions from the pkg node. So when visiting exprs within that class definitions non-error logging check can be true due to previous expr. Due to that we need to set that variable(non-error logging) to false before visiting those class definitions. Isn't it?

Reply:
If we are doing a check with `nonErrorLoggingCheck` set to true, until that check completes we don't/shouldn't set it to false. If we do, we are going against the expected behaviour. e.g., you set `nonErrorLoggingCheck` to true and do a compatibility check, but somewhere along the way if `nonErrorLoggingCheck` is changed to false, it will end up logging errors.

If you check the usage/assignments to `org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker#nonErrorLoggingCheck`, we don't explicitly set false to this anywhere (except in the changes in this PR), which guarantees the above.

 

> In here
> 
> https://github.com/ballerina-platform/ballerina-lang/blob/bab02a8fb3cfd11794e2394e1460b23ec0dfe115/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L8128
> 
> we are getting the class definitions from the pkg node. So when visiting exprs within that class definitions non-error logging check can be true due to previous expr. Due to that we need to set that variable(non-error logging) to false before visiting those class definitions. Isn't it?

I don't get why we need this. If someone sets `nonErrorLoggingCheck` to true they are expected to 
1. eventually do an error logging check
2. reset it to the previous state after the check (may be true or false)